components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/AuthorizedAPIManagementServiceImpl.java

```diff
@@ -0,0 +1,179 @@
+
+/**
+ * Authorized API management service implementation.
+ */
+public class AuthorizedAPIManagementServiceImpl implements AuthorizedAPIManagementService {
+
+    private final AuthorizedAPIDAO authorizedAPIDAO = new AuthorizedAPIDAOImpl();
+
+    @Override
+    public void addAuthorizedAPI(String applicationId, AuthorizedAPI authorizedAPI, String tenantDomain)
+            throws IdentityApplicationManagementException {
+
+        // Check if the application is a main application. If not, throw a client error.
+        ApplicationManagementService applicationManagementService = ApplicationManagementServiceImpl.getInstance();
+        String mainAppId = applicationManagementService.getMainAppId(applicationId);
+        if (mainAppId != null) {
+            throw buildClientException(INVALID_REQUEST, "Cannot add authorized APIs to a shared application.");
+        }
+        authorizedAPIDAO.addAuthorizedAPI(applicationId, authorizedAPI.getAPIId(),
+                authorizedAPI.getPolicyId(), authorizedAPI.getScopes(), IdentityTenantUtil.getTenantId(tenantDomain));
+    }
+
+    @Override
+    public void deleteAuthorizedAPIs(String appId, String apiId, String tenantDomain)
+            throws IdentityApplicationManagementException {
+
+
+        authorizedAPIDAO.deleteAuthorizedAPIs(appId, apiId, IdentityTenantUtil.getTenantId(tenantDomain));
+    }
+
+    @Override
+    public List<AuthorizedAPI> getAuthorizedAPIs(String applicationId, String tenantDomain)
+            throws IdentityApplicationManagementException {
+
+        try {
+            // Check if the application is a main application else get the main application id and main tenant id.
+            ApplicationManagementService applicationManagementService = ApplicationManagementServiceImpl.getInstance();
+            String mainAppId = applicationManagementService.getMainAppId(applicationId);
+            if (mainAppId != null) {
+                applicationId = mainAppId;
+                int tenantId = applicationManagementService.getTenantIdByApp(mainAppId);
+                tenantDomain = IdentityTenantUtil.getTenantDomain(tenantId);
+            }
+
+            List<AuthorizedAPI> authorizedAPIS = authorizedAPIDAO.getAuthorizedAPIs(applicationId,
+                    IdentityTenantUtil.getTenantId(tenantDomain));
+            for (AuthorizedAPI authorizedAPI : authorizedAPIS) {
+                // Get API resource data from DB.
+                APIResource apiResource = ApplicationManagementServiceComponentHolder.getInstance()
+                        .getAPIResourceManager().getAPIResourceById(authorizedAPI.getAPIId(), tenantDomain);
+                authorizedAPI.setAPIIdentifier(apiResource.getIdentifier());
+                authorizedAPI.setAPIName(apiResource.getName());
+                // Get Scope data from DB.
+                List<Scope> scopeList = new ArrayList<>();
+                for (Scope scope : authorizedAPI.getScopes()) {
```

Comment:
any possibility to return null from " authorizedAPI.getScopes()"? If yet, do a null check before looping